### ISSUE TITLE
viewer querier & styler : alternate WFS discovering

### ISSUE THREAD

**Issue by fphg** [georchestra/georchestra#1189]
The viewer's queryer and styler rely on a WFS related to the WMS service (ie a featureType related to a layer).
The actual implementation makes use of WFS 1.3.0 DescribeLayer operation. DescribeLayer describes how to reach the related WFS. DescribeLayer support isn't that commonly activated  (say deliberately deactivated !) beyond geOrchestra, thus querying and styling aren't supported that well.
INSPIRE solves this using discovery. A similar approach was implemented in the catalog panel.

To find a related WFS featureType given a WMS layer, you have to 
- search for a metadataURL (text/xml) in the WMS capabilities
- retrieve the dataset metadata
- identify the dataset id
- search for WFS operating on that dataset
- retrieve the WFS service metadata
- find the WFS locator
- query the WFS capabilities
- find the featureType

This seems a huge amount of work. On the other side, the true power of the viewer stands in its ability to combine services.

PS surprisingly some rare WMS do not allow DescribeLayer, yet allow custom SLD.

**Comment by fvanderbiest** [comment]
What a nightmare ...

**Comment by fphg** [comment]
Yep. Sounds weird. Actually it reflects the true data dispatch to services. You may serve a WMS, yet have the (multiple) WFS served on another platform/endpoint. With more and more datasets online we tends to import metadatas as is, 

It's not as nighmarish as you think.  First of all, we have a mechanism that can dig into remotes md stocks and find services endpoints. Very powerful because one can expand its SDI at no cost. Second, we partially implemented this to batch generate WFS/WCS metadatas and it worked surprisingly well.

Datadmins tends to duplicate datas as a workaround. This is definitively not scalable and generates other kinds of nightmares.

Before implementation proposal I'll try to prototype this.

**Comment by fxprunayre** [comment]
Related to that FYI, to avoid that "nighmarish" (and appropriate) search, some user are  using the following approach to "link" a dataset to WMS, WFS, WPS, ...: When encoding services in a dataset metadata create as many transfertOptions as services with one WMS and optionally one WFS, one WPS, .... You may have multiple groups of WMS/WFS/WPS per datasets. This is a shortcut made by Ifremer to enable in mapviewer some more advanced functionalities (eg. faceted search on features https://github.com/geonetwork/core-geonetwork/wiki/WFS-Filters-based-on-WFS-indexing-with-SOLR). 

![image](https://cloud.githubusercontent.com/assets/1701393/12477530/5e94f3e6-c030-11e5-88b2-433ea597cef4.png)

**Comment by fphg** [comment]
Thanks fx. This is convenient for our own datasets.
Unfortunately you can do this only if you're producer or delegate producer, ie you publish the dataset you produced. More on more datas come with their MD. We must import these MD as is (technically, and potentially legally). Customization occurs on the OWS endpoints and service metadatas, not on dataset metadatas.

Customizing MD isn't a good practice, you'll end with thousands of scan25 metadatas.

**Comment by fphg** [comment]
Some more info.
The french workgroup about services & co discussed this. Some wanted the so-called "strict mechanism" (discover OWS though CSW), some promoted the "soft" transferOption approach (multiple transfertOptions). We ended with following strategy on the service side :
* you MUST implement the strict scenario. Service metadatas are mandatory.
* you MAY implement the soft scenario if you want to promote your own services over the ones discovered though strict.

This caused last year bad evaluation of France because EU reporting was based on strict and ignored transferOptions.
See reports for strict http://geocat.docinspire.eu/tdb.php?db=std&harvest=geobretagne20160110&options=strict and soft http://geocat.docinspire.eu/tdb.php?db=std&harvest=geobretagne20160110

Downside, a client surely will find a service for each dataset, assuming the service MD were correctly harvested. It only may find a service if using the soft scenario.
